deepchem/models/torch_models/dtnn.py

```diff
@@ -0,0 +1,233 @@
+        n_distance: int (default 100)
+            granularity of distance matrix
+            step size will be (distance_max-distance_min)/n_distance
+        distance_min: float (default -1)
+            minimum distance of atom pairs (in Angstrom)
+        distance_max: float (default 18)
+            maximum distance of atom pairs (in Angstrom)
+        output_activation: bool (default True)
+            determines whether an activation function should be apply  to its output.
+        mode: str (default "regression")
+            Only "regression" is currently supported.
+        dropout: float (default 0.0)
+            the dropout probablity to use.
+        n_steps: int (default 2)
+            Number of DTNNStep Layers to use.
+
+        """
+        super(DTNN, self).__init__()
+        self.n_tasks = n_tasks
+        self.n_embedding = n_embedding
+        self.n_hidden = n_hidden
+        self.n_distance = n_distance
+        self.distance_min = distance_min
+        self.distance_max = distance_max
+        self.output_activation = output_activation
+        self.mode = mode
+        self.dropout = dropout
+        self.n_steps = n_steps
+
+        # get DTNNEmbedding
+        self.dtnn_embedding = layers.DTNNEmbedding(n_embedding=self.n_embedding)
+
+        # get DTNNSteps
+        self.dtnn_step = nn.ModuleList()
+        for i in range(self.n_steps):
+            self.dtnn_step.append(
+                layers.DTNNStep(n_embedding=self.n_embedding,
+                                n_distance=self.n_distance))
+
+        # get DTNNGather
+        self.dtnn_gather = layers.DTNNGather(
+            n_embedding=self.n_embedding,
+            layer_sizes=[self.n_hidden],
+            n_outputs=self.n_tasks,
+            output_activation=self.output_activation)
+
+        # get Final Linear Layer
+        self.linear = nn.LazyLinear(self.n_tasks)
+
+    def forward(self, inputs: List[torch.Tensor]):
+        """
+        Parameters
+        ----------
+        inputs: List
+            List of Tensors containing. (In the specified order)
+            - atom_number
+            - distance
+            - atom_membership
+            - distance_membership_i
+            - distance_membership_j
+
+        Returns
+        -------
+        output: torch.Tensor
+            Predictions of the Molecular Energy.
+
+        """
+        dtnn_embedding = self.dtnn_embedding(inputs[0])
+
+        for i in range(self.n_steps):
+            dtnn_embedding = nn.Dropout(self.dropout)(dtnn_embedding)
+            dtnn_embedding = self.dtnn_step[i](
+                [dtnn_embedding, inputs[1], inputs[3], inputs[4]])
+
+        dtnn_step = nn.Dropout(self.dropout)(dtnn_embedding)
+        dtnn_gather = self.dtnn_gather([dtnn_step, inputs[2]])
+
+        dtnn_gather = nn.Dropout(self.dropout)(dtnn_gather)
+        output = self.linear(dtnn_gather)
+        return output
+
+
+def _compute_features_on_batch(X_b: np.ndarray,
```

Comment:
Let's make this a public function and add an explicit unit test since it's quite complicated

Reply:
This should document it's designed to work with CoulombMatrix outputs

Reply:
For now, let's make a new file in utils for batching and add this there and document it

Reply:
Done